### PR TITLE
Add selectionExecuted and executionScheduled signals + update executed signal to include error status/info

### PR DESCRIPTION
**Pull request by mwakaba2**
## References

Issue #10259 

## Code changes

* Added 2 new signals: `executionScheduled` and `selectionExecuted`
  * `executionScheduled`: Emitted when a notebook cell execution got scheduled/started.
  *  `selectionExecuted`: Emitted after all selected notebook cells completed execution successfully/unsuccessfully. 
  * Note: @krassowski I didn't add the `executionStarted` signal because it didn't seem necessary. The started time had no noticeable difference compared to `executionScheduled`'s start time.
* Update the `executed` signal to include a `success` boolean and an optional `error` object.
* Added tests that checks whether the signals were accurately emitted. 

## User-facing changes

No visual changes. 

## Backwards-incompatible changes

Yes, there are two backwards-incompatible changes:

* If users rely on the old `Error('KernelReplyNotOK');` for error handling. I replaced this with a new error type: `KernelError(reply.content)` that takes in a `KernelMessage.IExecuteReplyMsg['content']` as its argument. 
* The `executed` signal's definition updated to emit `success` boolean + error object if an error occurs. 

## Items to address to complete this PR

- [x] meld `executionFailure` into the existing `executed` signal. The new `executed` signal will have a signature like the following:
```typescript
export const executed = new Signal<any, { notebook: Notebook; cell: Cell; success: boolean; error?: KernelError }>(
    {}
 );
```
- [x] the current `KernelError` implementation doesn't include the traceback, so I'll add that as well. 
- [x] update `allExecuted` naming to `selectionExecuted`.

### CODE REVIEW

**Comment by fcollonval** [comment]
Jumping a bit late in the discussion. But I got two suggestions:

- It would be better to use the existing signal `execute` for failure rather than introducing a new signal. Although it changes the current behavior, it is closer to what you would expect aka listening to one signal to know when the cell execution finishes. To figure out if the execution was successful, the object pass by the signal could return a boolean `success` and an `error` when success is False.
- I would choose another name than `allExecuted` because it may be confusing giving a false sense of `Run All cells` relation. Something like `selectionExecuted` seems more appropriate.

**Comment by krassowski** [comment]
Thank you @mwakaba2 for following through, and thanks @fcollonval for the comments.

> It would be better to use the existing signal execute for failure rather than introducing a new signal. Although it changes the current behavior, it is closer to what you would expect aka listening to one signal to know when the cell execution finishes. To figure out if the execution was successful, the object pass by the signal could return a boolean success and an error when success is False.

I'm fine with the suggestion to use the existing `execute` signal (I don't believe that any of the approaches is inherently better, just different pros & cons), but I still think that using proper errors (which can include extra information) is a nicer approach than using booleans.

Also agree with the suggestion on naming the other signal `selectionExecuted`.

**Comment by fcollonval** [comment]
> I still think that using proper errors (which can include extra information) is a nicer approach than using booleans (especially because in future it could include traceback and similar information).

Agree on that. The role of the boolean is to fulfill a strict contract, does it succeed? yes-no. Then if not you can look at other attributes (like the proposed `error`) to get more insight. But you are sure an error occurred even if `error` is falsy (`null` for example).

**Comment by mwakaba2** [comment]
@fcollonval @krassowski thank you for your feedback!

These are great suggestions. For the `execute` signal, I think it'll be helpful to have both the boolean and some additional attributes that'll inform the user more about the error. 

So to address your comments, I plan to make the following changes:

- meld `executionFailure` into the existing `executed` signal. The new `executed` signal will have a signature like the following:
```typescript
export const executed = new Signal<any, { notebook: Notebook; cell: Cell; success: boolean; error?: KernelError }>(
    {}
 );
```
- keep the `executionScheduled` signal that can be used to calculate code execution time instead of relying on recordTiming.
- the current `KernelError` implementation doesn't include the traceback, so I'll add that as well. 
-  update `allExecuted` naming to `selectionExecuted`. 

Does this sound good or did I miss something?

**Comment by krassowski** [comment]
Sounds good!

**Comment by mwakaba2** [comment]
Hi @krassowski and @fcollonval !
I addressed all your comments and would appreciate your second review!

File: packages/notebook/test/actions.spec.ts [hunk 1]
```diff
@@ -121,7 +121,7 @@ describe('@jupyterlab/notebook', () => {
       it('should emit an error for a cell execution failure.', async () => {
         let emitted = 0;
         let failed = 0;
-        let cellError = null;
+        let cellError: any = null;
```
Comment (krassowski) [inline packages/notebook/test/actions.spec.ts:124]: Could it be `cellError: KernelError | null` instead of `any`?

**Review by krassowski — COMMENTED** [review]
Thank you, looking great! Could you please rebase this PR and mark it as ready for review?

File: packages/notebook/src/actions.tsx [hunk 2]
```diff
@@ -43,6 +43,36 @@ import { nullTranslator, ITranslator } from '@jupyterlab/translation';
  */
 const JUPYTER_CELL_MIME = 'application/vnd.jupyter.cells';
 
+export class KernelError extends Error {
+  /**
+   * Exception name
+   */
+  errorName: string;
```
Comment (fcollonval) [inline packages/notebook/src/actions.tsx:50]:
```suggestion
  readonly errorName: string;
```

File: packages/notebook/src/actions.tsx [hunk 3]
```diff
@@ -43,6 +43,36 @@ import { nullTranslator, ITranslator } from '@jupyterlab/translation';
...
+  /**
+   * Exception value
+   */
+  errorValue: string;
```
Comment (fcollonval) [inline packages/notebook/src/actions.tsx:54]:
```suggestion
  readonly errorValue: string;
```

File: packages/notebook/src/actions.tsx [hunk 4]
```diff
@@ -43,6 +43,36 @@ import { nullTranslator, ITranslator } from '@jupyterlab/translation';
...
+  /**
+   * Traceback
+   */
+  traceback: string[];
```
Comment (fcollonval) [inline packages/notebook/src/actions.tsx:58]:
```suggestion
  readonly traceback: string[];
```

**Review by fcollonval — COMMENTED** [review]
Hey this looks almost good for me. In addition to @krassowski comment, I would be a bit more stringent on the error attributes visibility (see suggestions).

**Comment by mwakaba2** [comment]
thanks for the quick feedback! I'll make those changes, rebase, and mark this PR as ready for review. 🙂

**Review by fcollonval — APPROVED** [review]
Thanks for pushing this through @mwakaba2 

The CI errors are not related.

**Review by blink1073 — APPROVED** [review]
Thanks @mwakaba2!

File: packages/notebook/src/actions.tsx [hunk 5]
```diff
@@ -44,12 +74,40 @@ const JUPYTER_CELL_MIME = 'application/vnd.jupyter.cells';
...
+   */
+  static get executionScheduled(): ISignal<
+    any,
+    { notebook: Notebook; cell: Cell }
+  > {
+    return Private.executionScheduled;
+  }
+
+  /**
+   * A signal that emits whenever a cell completes execution.
```
Comment (krassowski) [inline packages/notebook/src/actions.tsx:102]: This docstring needs updating (copy-paste error). The docstring in implementation is fine.

Comment (krassowski) [reply packages/notebook/src/actions.tsx:102]: I opened https://github.com/jupyterlab/jupyterlab/pull/10579

**Pull request merged**